dask/bag/tests/test_bag.py

```diff
@@ -61,6 +61,30 @@ def test_keys():
     assert b.__dask_keys__() == sorted(dsk.keys())
 
 
+def test_bag_groupby_pure_hash(bagsequence):
+    """Test to ensure that `groupby` is grouping properly, when using 'pure' hashes
+    like integers. Eg., hash(False) == False, but hash('test') != 'test'. This is
+    testing the hash(False) == False case."""
+    assert isinstance(bagsequence, Bag)
+    # Probably a cleaner way to do this (maybe make them sets instead of lists??)
+    result = sorted(bagsequence.groupby(iseven).compute())
+    ordered_result = [(elem[0], sorted(elem[1])) for elem in result]
+    assert ordered_result == [(False, [1, 3, 5, 7, 9]), (True, [0, 2, 4, 6, 8])]
+
+
+def test_bag_groupby_normal_hash(bagsequence):
+    """Test to ensure that `groupby` is grouping properly, when using 'normal' hashes
+    like strings. Eg., hash(False) == False, but hash('test') != 'test'. This is
+    testing the hash('test') != 'test' case."""
+    ### THIS TEST CURRENTLY FAILS... THIS IS WHY IT IS BEING ADDED, TO HELP FIND
+    ### THE ISSUE AND KEEP TRACK OF IT.
+    assert isinstance(bagsequence, Bag)
+    # Probably a cleaner way to do this (maybe make them sets instead of lists??)
+    result = bagsequence.groupby(lambda x: "even" if (x % 2) == 0 else "odd").compute()
+    ordered_result = [(elem[0], sorted(elem[1])) for elem in sorted(result)]
+    assert ordered_result == [("even", [0, 2, 4, 6, 8]), ("odd", [1, 3, 5, 7, 9])]
```

Comment:
```suggestion
@pytest.mark.xfail(reason="https://github.com/dask/dask/issues/6640")
def test_bag_groupby_normal_hash():
    # https://github.com/dask/dask/issues/6640
    result = b.groupby(lambda x: "even" if iseven(x) else "odd").compute()
    assert len(result) == 2
    assert ("odd", [1, 3] * 3) in result
    assert ("even", [0, 2, 4] * 3) in result
```

Reply:
@jsignell Yes, please go ahead and make the changes, if you have the time!  Otherwise, I can.  But since I just don't know the Dask code very well (at all), I am happy to help but actually prefer someone like yourself puts the final blessing on the submission anyway.  😃

Reply:
Ok great! I'll take a look at what's failing...